### PR TITLE
Gather all test-related into test-mod.

### PR DESCRIPTION
**Pull request by ChengYueJia**
see more on #953

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit.rs
Comment (ChengYueJia) [inline zkevm-circuits/src/evm_circuit.rs:202]: This #1039 blocks gathering EvmCircuit test-related into test-mod.

File: zkevm-circuits/src/evm_circuit.rs
Comment (ChengYueJia) [inline zkevm-circuits/src/evm_circuit.rs:205]: This https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1039 blocks gathering EvmCircuit test-related into test-mod.

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit.rs:205]: Not sure I follow. What's the issue with this??

Comment (ChengYueJia) [reply zkevm-circuits/src/evm_circuit.rs:205]: > This #1039 blocks gathering EvmCircuit test-related into test-mod.

Apology for confusing you. I'll detail it.

For now, when we invoke `get_num_rows_required` of evm-circuit, we have to invoke  the `EvmCircuit::<F>::configure`. However, the `EvmCircuit::<F>::configure` is in `impl Circuit for EvmCircuit`. 
As #953 desc, we'd better hide the `impl Circuit for EvmCircuit` by moving it into test mod. 
So it's little block this pr.

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit.rs]: Ohh I see! Let's wait there and feel free to continue with this once it's merged!!!

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit.rs]: But where do we need to call `get_num_rows_required` that is not in the tests? I'm not sure if this is really blocking something.

Comment (ChengYueJia) [reply zkevm-circuits/src/evm_circuit.rs]: Oh.yeah. I got it(parm-face). 

I‘ve resolved the conflicts. Please review.

File: zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs
Comment (CPerezz) [inline zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs:881]: No sense to have imports local on each test. Please put all of them in common within the top level module imports.

File: zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs
Comment (CPerezz) [inline zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs:871]: Why isn't all that stuff moved to a test_util.rs file or whatever so that we reduce the line ammount and split a bit more the code?

**Review by CPerezz — CHANGES_REQUESTED** [review]
From the first review. A couple of comments that have to be addressed.

I think we should re-export the `XCircuit` in the `x_circuit.rs` file and we should make the modules non accessible. 

Therefore, you can access for example `CopyCircuit` since it's reexported in `copy_circuit.rs` but no access to `copy_circuit::tests::CopyCircuit` is needed.
Same applies for all other circuits.


I've also seen that there are a lot of `dev.rs` files which have been removed but were serving the specific purpose that it's intended to be solved by this PR.
The comment made for the `ByteCode` circuit related to the `test_util.rs` file applies for all other circuits.

File: zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs
Comment (ChengYueJia) [reply zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs:881]: This is because moving import outside the method will meet `unused import` error when using make clippy.
As below.
image

File: zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs
Comment (ChengYueJia) [reply zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs:871]: Make sense.

Comment (ChengYueJia) [reply zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs]: I've separated test-mod into per file.

**Comment by ChengYueJia** [comment]
> I think we should re-export the `XCircuit` in the `x_circuit.rs` file and we should make the modules non accessible.

I've re-exported them as `pub use test::XxxCircuit as TestXxxCircuit` . 

> I've also seen that there are a lot of `dev.rs` files which have been removed but were serving the specific purpose that it's intended to be solved by this PR. The comment made for the `ByteCode` circuit related to the `test_util.rs` file applies for all other circuits.

I've separated test-mod into per file in`xxx_circuit/test.rs`.

@CPerezz Please have another review.

File: zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs
Comment (ChengYueJia) [reply zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs]: > This is because moving import outside the method will meet `unused import` error when using make clippy. (As below.)

I resolve this by adding `#![allow(unused_imports)]` to the mod.

It's a little weird.

**Comment by aguzmant103** [comment]
@ChengYueJia sorry for the long time to review your latest changes.

Can you resolve the latest conflicts?  Afterwards @leolara can review it so we can merge this

**Comment by ChengYueJia** [comment]
> Can you resolve the latest conflicts? Afterwards @leolara can review it so we can merge this

I'll resolve this one first. When I finished, I'll let you know.

**Comment by ChengYueJia** [comment]
> @ChengYueJia I guess we should close this in favor of #1300 ?

yep, I'll close this.

**Pull request closed without merging**
